node/common/client_service_base.ts

```diff
@@ -0,0 +1,172 @@
+
+/**
+ * This class handles all client interactions including registering certificates
+ * and contracts, listing contracts, validating the ledger, and executing
+ * contracts.
+ */
+export class ClientServiceBase {
+  readonly metadata: Metadata;
+  readonly ledgerClient: LedgerClient;
+  readonly ledgerPrivileged: LedgerPrivileged;
+  readonly auditorClient: AuditorClient;
+  readonly auditorPrivileged: AuditorPrivileged;
+  readonly signerFactory: SignatureSignerFactory;
+  readonly protobuf: ScalarMessage;
+  readonly onBrowser: boolean;
+  readonly config: ClientConfig;
+
+  signer: SignatureSigner | null = null;
+
+  /**
+   * @param {ScalarService & {signerFactory: SignatureSignerFactory}} services contains the object of ledgeClient and
+   *  the object of ledgerPrivileged
+   * @param {ScalarMessage} protobuf protobuf object to inject
+   * @param {Properties} properties JSON Object used for setting client properties
+   * @param {Metadata} metadata gRPC metadata object used to add header
+   * @param {boolean} [onBrowser=false] whether the client is running on browser or not
+   *  to the gRPC request
+   */
+  constructor(
+    services: ScalarService & {
+      signerFactory: SignatureSignerFactory;
+    },
+    protobuf: ScalarMessage,
+    properties: Properties,
+    metadata: Metadata,
+    // eslint-disable-next-line @typescript-eslint/no-inferrable-types
+    onBrowser: boolean = false
+  ) {
+    this.config = new ClientConfig(properties);
+    this.metadata = metadata;
+    this.ledgerClient = services['ledgerClient'];
+    this.ledgerPrivileged = services['ledgerPrivileged'];
+    this.auditorClient = services['auditorClient'];
+    this.auditorPrivileged = services['auditorPrivileged'];
+    this.signerFactory = services['signerFactory'];
+    this.protobuf = protobuf;
+    this.onBrowser = onBrowser;
+  }
+
+  static get binaryStatusKey(): string {
+    return 'rpc.status-bin';
+  }
+
+  /**
+   * Registers the certificate specified in the given client properties.
+   * @throws {ClientError|Error}
+   */
+  async registerCertificate(): Promise<void> {
+    const request = await this.createCertificateRegistrationRequest();
+
+    if (this.isAuditorEnabled()) {
+      const promise = new Promise<void>((resolve, reject) => {
+        this.auditorPrivileged.registerCert(request, this.metadata, err => {
+          if (err) reject(err);
+          else resolve();
+        });
+      });
+
+      await this.executePromise(promise);
+    }
+
+    const promise = new Promise<void>((resolve, reject) => {
+      this.ledgerPrivileged.registerCert(request, this.metadata, err => {
+        if (err) reject(err);
+        else resolve();
+      });
+    });
+
+    return this.executePromise(promise) as Promise<void>;
+  }
+
+  /**
+   * Create the byte array of CertificateRegistrationRequest.
+   * @return {Promise<Uint8Array>}
+   * @throws {Error}
+   */
+  async createSerializedCertificateRegistrationRequest(): Promise<Uint8Array> {
```

Comment:
Just a quick question. Will registerCertificate() with serialized binary created by this function be provided with another upcoming PR?

Reply:
@jnmt 

Thank you for your question.

Yes. There will be a ClientServiceWithBinary class in Node part that will implement it.
Like https://github.com/scalar-labs/scalardl-node-client-sdk/blob/master/scalardl-node-client-sdk.js#L175

Because the ClientServiceWithBinary will be implemented for Node but not for Browser so I didn't put it in the common part.

Reply:
> Because the ClientServiceWithBinary will be implemented for Node but not for Browser so I didn't put it in the common part.

Ahh, thanks!